### ISSUE TITLE
Type issue with propertyConverters()

### ISSUE THREAD

**Issue by davidwu226** [evermeer/EVReflection#48]
```
class y : EVObject {
    var id = "y id"
}

class z: EVObject {
    var foo: String? = "foo"
    var bar: String? = "bar"
    var arr: [y] = []

    override func propertyConverters() -> [(String?, (Any?) -> (), () -> Any?)] {
        return [
            ( "arr",
                { self.arr = $0 as! [y] },
                { return self.arr.map({$0.id}) }
            )
        ]
    }  
}

let json = "{\"foo\": \"foobar\", \"arr\": [{\"id\": \"test\"}, {\"id\": \"test2\"}]}"
let r = z(json: json)
```

The above fails I think because the propertyConverter getter here returns a different type from the setter (this is the case where I want to output id's for subobjects when serializing to JSON but want to read full subobjects when deserializing from JSON).

I think the fix is this patch below -- basically, ignore the type from propertyConverter. However, I'm not sure if it's correct:

```
diff --git a/EVReflection/pod/EVReflection.swift b/EVReflection/pod/EVReflection.swift
index e2fc6d0..ff06139 100644
--- a/EVReflection/pod/EVReflection.swift
+++ b/EVReflection/pod/EVReflection.swift
@@ -852,12 +855,15 @@ final public class EVReflection {
                 }
                 if !skipThisKey {
                     var value = property.value
+                    // Convert the Any value to a NSObject value
+                    var (unboxedValue, valueType, isObject) = valueForAny(theObject, key: originalKey, anyValue: value)
+
                     // If there is a properyConverter, then use the result of that instead.
                     if let (_, _, propertyGetter) = (theObject as? EVObject)?.propertyConverters().filter({$0.0 == originalKey}).first {
                         value = propertyGetter()
+                        // Convert the Any value to a NSObject value but use original type.
+                        (unboxedValue, _, isObject) = valueForAny(theObject, key: originalKey, anyValue: value)
                     }
-                    // Convert the Any value to a NSObject value
-                    var (unboxedValue, valueType, isObject) = valueForAny(theObject, key: originalKey, anyValue: value)
                     if isObject {
                         // sub objects will be added as a dictionary itself.
                         let (dict, _) = toDictionary(unboxedValue as! NSObject, performKeyCleanup: performKeyCleanup)
```

**Comment by evermeer** [comment]
Before I dive into this let me first try to explain how I think this should work:
In your case your arr is an array of objects. When the json is parsed it will first be converted to arrays and dictionaries. Here the setter will be called with an array of dictionaries. That array of dictionaries will not be the same as the array of y objects. 

so in order to make the setter work it should be changed to something like (untested)
self.arr = $0.map({y($0["id"])})
Then you y object would also need a constructor that takes an id to create itself.

I will now have a look at your code to see if you have a valid point.

**Comment by evermeer** [comment]
Ah! I see what you are doing. You first call the valueForAny to convert the object before it's forwarded to the propertyGetter. That indeed would make it easier in the propertyGetter. Let me think about this for a moment and let me do some tests to see if this is indeed what someone usually wants and if there are no side effects. Are you sure you have to call the valueForAny again after the call to the propertyGetter? the results are within the if and not used anywhere. 

When I implement your change then the existing testCustomPropertyConverter unit test will fail. If you do have code where all unit tests are successful, maybe even with an added unit test of your issue, then that would be great if you could do a pull request for that.

By the way... Great to see that you are diving into the code!

**Comment by davidwu226** [comment]
I realized my contrived example wasn't very good. Imagine class y is more like defined as follows:

```
class y : EVObject {
    var id = "y id"
    var name = "some name"
    var price = 12.30
    var phone = "555-5555"
}
```

Now, when class y is embedded in class z's "arr" array, when serializing to JSON, the intended behavior I wanted is to output that as {"arr": [ "y.id1", "y,id2", ...]}. However, when deserializing from JSON, I expect the JSON string to be {"arr": [ {"id": "y.id1", "name": "bob", "price": "500.00", "phone": "666-6666"}, ...]}. So my intention is that the types for serialization (array of Strings) and de-serialization (array of class y) does not match.

So the issue is that on serializing, the above code gets called to get the type -- but since the new getter type is no longer a array of class y but a array of String, the serialization fails.

The patch above addresses by getting the type before calling the new getter.

I think the second valueForAny is needed -- that's needed for the case where you are calling this for de-serialization.

**Comment by evermeer** [comment]
A couple of other small changes where needed to make this work correctly. I just pushed a fix as version 2.19.1 to GitHub and cocoapods.
